File: packages/modules/lodash/core/go/core/lang/lang.go
package lang

import (
	"fmt"
	"math"
	"reflect"
	"regexp"
	"time"
)

func Gt(value float64, other float64) bool {
	return value > other
}

func Gte(value float64, other float64) bool {
	return value >= other
}

func Lt(value float64, other float64) bool {
	return value < other
}

func Lte(value float64, other float64) bool {
	return value <= other
}

func CastArray(value interface{}) []interface{} {
	if value == nil {
		return []interface{}{}
	}
	val := reflect.ValueOf(value)
	if val.Kind() == reflect.Slice || val.Kind() == reflect.Array {
		result := make([]interface{}, val.Len())
		for i := 0; i < val.Len(); i++ {
			result[i] = val.Index(i).Interface()
		}
		return result
	}
	return []interface{}{value}
}

func Clone(value interface{}) interface{} {
	if value == nil {
		return nil
	}
	val := reflect.ValueOf(value)
	switch val.Kind() {
	case reflect.Slice:
		result := reflect.MakeSlice(val.Type(), val.Len(), val.Cap())
		reflect.Copy(result, val)
		return result.Interface()
	case reflect.Map:
		result := reflect.MakeMap(val.Type())
		for _, key := range val.MapKeys() {
			result.SetMapIndex(key, val.MapIndex(key))
		}
		return result.Interface()
	case reflect.Ptr:
		if val.IsNil() {
			return nil
		}
		elem := val.Elem()
		newPtr := reflect.New(elem.Type())
		newPtr.Elem().Set(elem)
		return newPtr.Interface()
	case reflect.Struct:
		cp := reflect.New(val.Type()).Elem()
		cp.Set(val)
		return cp.Interface()
	default:
		return value
	}
}

func cloneDeepReflect(val reflect.Value) reflect.Value {
	switch val.Kind() {
	case reflect.Ptr:
		if val.IsNil() {
			return reflect.Zero(val.Type())
		}
		newPtr := reflect.New(val.Type().Elem())
		newPtr.Elem().Set(cloneDeepReflect(val.Elem()))
		return newPtr
	case reflect.Struct:
		cp := reflect.New(val.Type()).Elem()
		for i := 0; i < val.NumField(); i++ {
			cp.Field(i).Set(cloneDeepReflect(val.Field(i)))
		}
		return cp
	case reflect.Map:
		if val.IsNil() {
			return reflect.Zero(val.Type())
		}
		cp := reflect.MakeMap(val.Type())
		for _, key := range val.MapKeys() {
			cp.SetMapIndex(cloneDeepReflect(key), cloneDeepReflect(val.MapIndex(key)))
		}
		return cp
	case reflect.Slice:
		if val.IsNil() {
			return reflect.Zero(val.Type())
		}
		cp := reflect.MakeSlice(val.Type(), val.Len(), val.Cap())
		for i := 0; i < val.Len(); i++ {
			cp.Index(i).Set(cloneDeepReflect(val.Index(i)))
		}
		return cp
	case reflect.Array:
		cp := reflect.New(val.Type()).Elem()
		for i := 0; i < val.Len(); i++ {
			cp.Index(i).Set(cloneDeepReflect(val.Index(i)))
		}
		return cp
	case reflect.Interface:
		if val.IsNil() {
			return reflect.Zero(val.Type())
		}
		return cloneDeepReflect(val.Elem())
	default:
		return val
	}
}

func CloneDeep(value interface{}) interface{} {
	if value == nil {
		return nil
	}
	val := reflect.ValueOf(value)
	result := cloneDeepReflect(val)
	return result.Interface()
}

func CloneDeepWith(value interface{}, customizer func(interface{}) interface{}) interface{} {
	if customizer != nil {
		if result := customizer(value); result != nil {
			return result
		}
	}
	return CloneDeep(value)
}

func CloneWith(value interface{}, customizer func(interface{}) interface{}) interface{} {
	if customizer != nil {
		if result := customizer(value); result != nil {
			return result
		}
	}
	return Clone(value)
}

func ConformsTo(object interface{}, source map[string]func(interface{}) bool) bool {
	val := reflect.ValueOf(object)
	if val.Kind() == reflect.Ptr {
		val = val.Elem()
	}
	if !val.IsValid() {
		return false
	}
	for key, predicate := range source {
		var propVal interface{}
		if val.Kind() == reflect.Map {
			mapVal := val.MapIndex(reflect.ValueOf(key))
			if !mapVal.IsValid() {
				return false
			}
			propVal = mapVal.Interface()
		} else if val.Kind() == reflect.Struct {
			field := val.FieldByName(key)
			if !field.IsValid() {
				return false
			}
			propVal = field.Interface()
		} else {
			return false
		}
		if !predicate(propVal) {
			return false
		}
	}
	return true
}

func Eq(value interface{}, other interface{}) bool {
	return reflect.DeepEqual(value, other)
}

func IsArguments(value interface{}) bool {
	if value == nil {
		return false
	}
	return reflect.ValueOf(value).Kind() == reflect.Array
}

func IsArray(value interface{}) bool {
	if value == nil {
		return false
	}
	k := reflect.ValueOf(value).Kind()
	return k == reflect.Array || k == reflect.Slice
}

func IsArrayBuffer(value interface{}) bool {
	if value == nil {
		return false
	}
	v := reflect.ValueOf(value)
	return v.Kind() == reflect.Slice && v.Type().Elem().Kind() == reflect.Uint8
}

func IsArrayLike(value interface{}) bool {
	if value == nil {
		return false
	}
	switch reflect.ValueOf(value).Kind() {
	case reflect.Slice, reflect.Array, reflect.String, reflect.Map:
		return true
	}
	return false
}

func IsArrayLikeObject(value interface{}) bool {
	if value == nil {
		return false
	}
	k := reflect.ValueOf(value).Kind()
	return k == reflect.Slice || k == reflect.Array || k == reflect.Map
}

func IsBoolean(value interface{}) bool {
	if value == nil {
		return false
	}
	_, ok := value.(bool)
	return ok
}

func IsBuffer(value interface{}) bool {
	return false
}

func IsDate(value interface{}) bool {
	if value == nil {
		return false
	}
	_, ok := value.(time.Time)
	return ok
}

func IsElement(value interface{}) bool {
	return false
}

func IsEmpty(value interface{}) bool {
	if value == nil {
		return true
	}
	val := reflect.ValueOf(value)
	switch val.Kind() {
	case reflect.String, reflect.Slice, reflect.Array, reflect.Map:
		return val.Len() == 0
	case reflect.Struct:
		return reflect.DeepEqual(value, reflect.Zero(val.Type()).Interface())
	case reflect.Ptr, reflect.Interface:
		if val.IsNil() {
			return true
		}
		return IsEmpty(val.Elem().Interface())
	case reflect.Func, reflect.Chan:
		return val.IsNil()
	default:
		return false
	}
}

func IsEqual(value interface{}, other interface{}) bool {
	return reflect.DeepEqual(value, other)
}

func IsEqualWith(value interface{}, other interface{}, customizer func(interface{}, interface{}) bool) bool {
	if customizer != nil {
		return customizer(value, other)
	}
	return reflect.DeepEqual(value, other)
}

func IsError(value interface{}) bool {
	if value == nil {
		return false
	}
	_, ok := value.(error)
	return ok
}

func IsFinite(value float64) bool {
	return !math.IsInf(value, 0) && !math.IsNaN(value)
}

func IsFunction(value interface{}) bool {
	if value == nil {
		return false
	}
	return reflect.ValueOf(value).Kind() == reflect.Func
}

func IsInteger(value float64) bool {
	if math.IsInf(value, 0) || math.IsNaN(value) {
		return false
	}
	return value == float64(int64(value))
}

func IsLength(value int) bool {
	return value >= 0 && value <= math.MaxUint32
}

func IsMap(value interface{}) bool {
	if value == nil {
		return false
	}
	return reflect.ValueOf(value).Kind() == reflect.Map
}

func IsMatch(object interface{}, source interface{}) bool {
	if source == nil {
		return true
	}
	if object == nil {
		return false
	}
	srcVal := reflect.ValueOf(source)
	objVal := reflect.ValueOf(object)
	if srcVal.Kind() == reflect.Ptr {
		srcVal = srcVal.Elem()
	}
	if objVal.Kind() == reflect.Ptr {
		objVal = objVal.Elem()
	}
	if !srcVal.IsValid() || !objVal.IsValid() {
		return false
	}
	switch srcVal.Kind() {
	case reflect.Map:
		if objVal.Kind() != reflect.Map {
			return false
		}
		for _, key := range srcVal.MapKeys() {
			objMapVal := objVal.MapIndex(key)
			if !objMapVal.IsValid() {
				return false
			}
			if !IsMatch(objMapVal.Interface(), srcVal.MapIndex(key).Interface()) {
				return false
			}
		}
		return true
	case reflect.Struct:
		if objVal.Kind() != reflect.Struct {
			return false
		}
		for i := 0; i < srcVal.NumField(); i++ {
			if !srcVal.Type().Field(i).IsExported() {
				continue
			}
			name := srcVal.Type().Field(i).Name
			objField := objVal.FieldByName(name)
			if !objField.IsValid() {
				return false
			}
			if !IsMatch(objField.Interface(), srcVal.Field(i).Interface()) {
				return false
			}
		}
		return true
	case reflect.Slice, reflect.Array:
		if objVal.Kind() != reflect.Slice && objVal.Kind() != reflect.Array {
			return false
		}
		if srcVal.Len() != objVal.Len() {
			return false
		}
		for i := 0; i < srcVal.Len(); i++ {
			if !IsMatch(objVal.Index(i).Interface(), srcVal.Index(i).Interface()) {
				return false
			}
		}
		return true
	default:
		return reflect.DeepEqual(object, source)
	}
}

func IsMatchWith(object interface{}, source interface{}, customizer func(interface{}, interface{}) bool) bool {
	if customizer != nil {
		return customizer(object, source)
	}
	return IsMatch(object, source)
}

func IsNaN(value interface{}) bool {
	if value == nil {
		return false
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Float32, reflect.Float64:
		return math.IsNaN(v.Float())
	case reflect.Interface:
		return IsNaN(v.Elem().Interface())
	}
	return false
}

func IsNative(value interface{}) bool {
	return false
}

func IsNil(value interface{}) bool {
	if value == nil {
		return true
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Ptr, reflect.Interface, reflect.Slice, reflect.Map, reflect.Chan, reflect.Func:
		return v.IsNil()
	}
	return false
}

func IsNull(value interface{}) bool {
	return IsNil(value)
}

func IsNumber(value interface{}) bool {
	if value == nil {
		return false
	}
	switch reflect.ValueOf(value).Kind() {
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
		reflect.Float32, reflect.Float64,
		reflect.Complex64, reflect.Complex128:
		return true
	}
	return false
}

func IsObject(value interface{}) bool {
	if value == nil {
		return false
	}
	switch reflect.ValueOf(value).Kind() {
	case reflect.Struct, reflect.Map, reflect.Slice, reflect.Array,
		reflect.Ptr, reflect.Func, reflect.Chan, reflect.Interface:
		return true
	}
	return false
}

func IsObjectLike(value interface{}) bool {
	if value == nil {
		return false
	}
	k := reflect.ValueOf(value).Kind()
	return k != reflect.Invalid && k != reflect.Bool && k != reflect.String &&
		k != reflect.Int && k != reflect.Int8 && k != reflect.Int16 &&
		k != reflect.Int32 && k != reflect.Int64 &&
		k != reflect.Uint && k != reflect.Uint8 && k != reflect.Uint16 &&
		k != reflect.Uint32 && k != reflect.Uint64 &&
		k != reflect.Float32 && k != reflect.Float64 &&
		k != reflect.Complex64 && k != reflect.Complex128
}

func IsPlainObject(value interface{}) bool {
	if value == nil {
		return false
	}
	v := reflect.ValueOf(value)
	if v.Kind() == reflect.Ptr {
		v = v.Elem()
	}
	return v.Kind() == reflect.Map || v.Kind() == reflect.Struct
}

func IsRegExp(value interface{}) bool {
	if value == nil {
		return false
	}
	_, ok := value.(*regexp.Regexp)
	return ok
}

func IsSafeInteger(value float64) bool {
	if math.IsInf(value, 0) || math.IsNaN(value) {
		return false
	}
	if value != float64(int64(value)) {
		return false
	}
	maxSafe := math.Pow(2, 53) - 1
	return value >= -maxSafe && value <= maxSafe
}

func IsSet(value interface{}) bool {
	if value == nil {
		return false
	}
	v := reflect.ValueOf(value)
	if v.Kind() != reflect.Map {
		return false
	}
	var emptyStruct struct{}
	return v.Type().Elem() == reflect.TypeOf(emptyStruct)
}

func IsString(value interface{}) bool {
	if value == nil {
		return false
	}
	_, ok := value.(string)
	return ok
}

func IsSymbol(value interface{}) bool {
	return false
}

func IsTypedArray(value interface{}) bool {
	if value == nil {
		return false
	}
	v := reflect.ValueOf(value)
	if v.Kind() != reflect.Slice && v.Kind() != reflect.Array {
		return false
	}
	switch v.Type().Elem().Kind() {
	case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
		reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64,
		reflect.Float32, reflect.Float64, reflect.Uintptr:
		return true
	}
	return false
}

func IsUndefined(value interface{}) bool {
	return value == nil
}

func IsWeakMap(value interface{}) bool {
	return false
}

func IsWeakSet(value interface{}) bool {
	return false
}

func ToArray(value interface{}) []interface{} {
	if value == nil {
		return []interface{}{}
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Slice, reflect.Array:
		result := make([]interface{}, v.Len())
		for i := 0; i < v.Len(); i++ {
			result[i] = v.Index(i).Interface()
		}
		return result
	case reflect.Map:
		result := make([]interface{}, 0, v.Len())
		for _, key := range v.MapKeys() {
			result = append(result, v.MapIndex(key).Interface())
		}
		return result
	case reflect.String:
		s := v.String()
		result := make([]interface{}, 0, len(s))
		for _, r := range s {
			result = append(result, string(r))
		}
		return result
	default:
		return []interface{}{value}
	}
}

func ToFinite(value interface{}) float64 {
	if value == nil {
		return 0
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Float32, reflect.Float64:
		f := v.Float()
		if math.IsNaN(f) {
			return 0
		}
		if math.IsInf(f, 1) {
			return math.MaxFloat64
		}
		if math.IsInf(f, -1) {
			return -math.MaxFloat64
		}
		return f
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return float64(v.Int())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return float64(v.Uint())
	case reflect.Bool:
		if v.Bool() {
			return 1
		}
		return 0
	case reflect.String:
		var f float64
		fmt.Sscanf(v.String(), "%f", &f)
		return f
	default:
		return 0
	}
}

func ToInteger(value interface{}) int {
	if value == nil {
		return 0
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Float32, reflect.Float64:
		f := v.Float()
		if math.IsNaN(f) || math.IsInf(f, 0) {
			return 0
		}
		return int(f)
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return int(v.Int())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return int(v.Uint())
	case reflect.Bool:
		if v.Bool() {
			return 1
		}
		return 0
	case reflect.String:
		var i int
		fmt.Sscanf(v.String(), "%d", &i)
		return i
	default:
		return 0
	}
}

func ToLength(value interface{}) int {
	n := ToInteger(value)
	if n < 0 {
		return 0
	}
	if n > math.MaxUint32 {
		return math.MaxUint32
	}
	return n
}

func ToNumber(value interface{}) float64 {
	if value == nil {
		return 0
	}
	v := reflect.ValueOf(value)
	switch v.Kind() {
	case reflect.Float32, reflect.Float64:
		return v.Float()
	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
		return float64(v.Int())
	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
		return float64(v.Uint())
	case reflect.Bool:
		if v.Bool() {
			return 1
		}
		return 0
	case reflect.String:
		var f float64
		fmt.Sscanf(v.String(), "%f", &f)
		return f
	default:
		return 0
	}
}

func ToPlainObject(value interface{}) map[string]interface{} {
	result := make(map[string]interface{})
	if value == nil {
		return result
	}
	v := reflect.ValueOf(value)
	if v.Kind() == reflect.Ptr {
		v = v.Elem()
	}
	if !v.IsValid() {
		return result
	}
	switch v.Kind() {
	case reflect.Struct:
		t := v.Type()
		for i := 0; i < v.NumField(); i++ {
			field := t.Field(i)
			if field.IsExported() {
				result[field.Name] = v.Field(i).Interface()
			}
		}
	case reflect.Map:
		for _, key := range v.MapKeys() {
			result[fmt.Sprint(key.Interface())] = v.MapIndex(key).Interface()
		}
	}
	return result
}

func ToSafeInteger(value interface{}) int {
	n := ToInteger(value)
	if n < 0 {
		return 0
	}
	maxSafe := int(math.Pow(2, 53) - 1)
	if n > maxSafe {
		return maxSafe
	}
	return n
}

func ToString(value interface{}) string {
	if value == nil {
		return ""
	}
	return fmt.Sprint(value)
}

File: packages/modules/lodash/core/rust/src/core/lang.rs
use std::any::Any;
use std::collections::HashMap;
use std::fmt::Display;
use std::time::SystemTime;

/// Wraps a value in a Vec.
pub fn cast_array<T>(value: T) -> Vec<T> {
    vec![value]
}

/// Creates a shallow clone of value.
pub fn clone<T: Clone>(value: &T) -> T {
    value.clone()
}

/// Creates a deep clone of value.
pub fn clone_deep<T: Clone>(value: &T) -> T {
    value.clone()
}

/// Checks if object conforms to source by comparing all source properties.
pub fn conforms_to<K: Eq + std::hash::Hash, V: PartialEq>(
    object: &HashMap<K, V>,
    source: &HashMap<K, V>,
) -> bool {
    source.iter().all(|(k, v)| object.get(k) == Some(v))
}

/// Checks if value is the same value as other (SameValueZero equality).
pub fn eq<T: PartialEq>(value: &T, other: &T) -> bool {
    value == other
}

/// Checks if value is greater than other.
pub fn gt<T: PartialOrd>(value: T, other: T) -> bool {
    value > other
}

/// Checks if value is greater than or equal to other.
pub fn gte<T: PartialOrd>(value: T, other: T) -> bool {
    value >= other
}

/// Checks if value is array-like (a slice reference).
pub fn is_array<T>(_value: &[T]) -> bool {
    true
}

/// Checks if value is a boolean.
pub fn is_boolean(value: &dyn Any) -> bool {
    value.is::<bool>()
}

/// Checks if value is a date (SystemTime).
pub fn is_date(value: &dyn Any) -> bool {
    value.is::<SystemTime>()
}

/// Checks if value is empty (empty slice).
pub fn is_empty<T>(value: &[T]) -> bool {
    value.is_empty()
}

/// Checks if values are equivalent (SameValueZero).
pub fn is_equal<T: PartialEq>(value: &T, other: &T) -> bool {
    value == other
}

/// Checks if values are equivalent using a custom comparator.
pub fn is_equal_with<T, F>(value: &T, other: &T, comparator: F) -> bool
where
    F: Fn(&T, &T) -> bool,
{
    comparator(value, other)
}

/// Checks if value is an error.
pub fn is_error(value: &dyn Any) -> bool {
    value.is::<String>() || value.is::<std::io::Error>()
}

/// Checks if value is a finite number.
pub fn is_finite(value: f64) -> bool {
    value.is_finite()
}

/// Checks if value is a function (always false in Rust).
pub fn is_function() -> bool {
    false
}

/// Checks if value is an integer.
pub fn is_integer(value: f64) -> bool {
    value.fract() == 0.0 && value.is_finite()
}

/// Checks if value is a valid array-like length (0 to 2^32-1).
pub fn is_length(value: usize) -> bool {
    value <= u32::MAX as usize
}

/// Checks if value is a HashMap.
pub fn is_map<K, V>(_value: &HashMap<K, V>) -> bool {
    true
}

/// Checks if value is NaN.
pub fn is_nan(value: f64) -> bool {
    value.is_nan()
}

/// Checks if value is None.
pub fn is_nil<T>(value: &Option<T>) -> bool {
    value.is_none()
}

/// Checks if value is None.
pub fn is_null<T>(value: &Option<T>) -> bool {
    value.is_none()
}

/// Checks if value is a number.
pub fn is_number(value: &dyn Any) -> bool {
    value.is::<i8>()
        || value.is::<i16>()
        || value.is::<i32>()
        || value.is::<i64>()
        || value.is::<u8>()
        || value.is::<u16>()
        || value.is::<u32>()
        || value.is::<u64>()
        || value.is::<f32>()
        || value.is::<f64>()
        || value.is::<isize>()
        || value.is::<usize>()
}

/// Checks if value is a number (always true for f64).
pub fn is_number_f64(_value: f64) -> bool {
    true
}

/// Checks if value is object-like.
pub fn is_object<T>(_value: &T) -> bool {
    true
}

/// Checks if value is a plain object.
pub fn is_plain_object<T>(_value: &T) -> bool {
    true
}

/// Checks if value is a RegExp (always false without regex crate).
pub fn is_reg_exp(_value: &dyn Any) -> bool {
    false
}

/// Checks if value is a safe integer.
pub fn is_safe_integer(value: f64) -> bool {
    let max_safe = 9007199254740991.0;
    value.is_finite() && value.fract() == 0.0 && value >= -max_safe && value <= max_safe
}

/// Checks if value is a Set (always false).
pub fn is_set() -> bool {
    false
}

/// Checks if value is a String.
pub fn is_string(value: &dyn Any) -> bool {
    value.is::<String>()
}

/// Checks if value is a Symbol (always false).
pub fn is_symbol() -> bool {
    false
}

/// Checks if value is a typed array (slice reference).
pub fn is_typed_array<T>(_value: &[T]) -> bool {
    true
}

/// Checks if value is undefined (always false in Rust).
pub fn is_undefined() -> bool {
    false
}

/// Checks if value is a WeakMap (always false).
pub fn is_weak_map() -> bool {
    false
}

/// Checks if value is a WeakSet (always false).
pub fn is_weak_set() -> bool {
    false
}

/// Checks if value is less than other.
pub fn lt<T: PartialOrd>(value: T, other: T) -> bool {
    value < other
}

/// Checks if value is less than or equal to other.
pub fn lte<T: PartialOrd>(value: T, other: T) -> bool {
    value <= other
}

/// Converts value to a Vec.
pub fn to_array<T: Clone>(value: &[T]) -> Vec<T> {
    value.to_vec()
}

/// Converts value to a finite number.
pub fn to_finite(value: f64) -> f64 {
    if value.is_finite() {
        value
    } else {
        0.0
    }
}

/// Converts value to an integer.
pub fn to_integer(value: f64) -> i64 {
    value as i64
}

/// Converts value to a valid array-like length (capped at 2^32-1).
pub fn to_length(value: usize) -> usize {
    value.min(u32::MAX as usize)
}

/// Converts a string to a number.
pub fn to_number(value: &str) -> f64 {
    value.parse().unwrap_or(0.0)
}

/// Returns the value as-is (identity function).
pub fn to_plain_object<T>(value: T) -> T {
    value
}

/// Converts value to a safe integer.
pub fn to_safe_integer(value: f64) -> i64 {
    let max_safe: i64 = 9007199254740991;
    let i = value as i64;
    if i >= -max_safe && i <= max_safe {
        i
    } else {
        0
    }
}

/// Converts a value to its string representation.
pub fn to_string(value: &dyn Display) -> String {
    format!("{}", value)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cast_array() {
        assert_eq!(cast_array(42), vec![42]);
    }

    #[test]
    fn test_clone_and_clone_deep() {
        let v = vec![1, 2, 3];
        assert_eq!(clone(&v), v);
        assert_eq!(clone_deep(&v), v);
    }

    #[test]
    fn test_conforms_to() {
        let mut obj = HashMap::new();
        obj.insert("a".to_string(), 1);
        obj.insert("b".to_string(), 2);
        let mut source = HashMap::new();
        source.insert("a".to_string(), 1);
        assert!(conforms_to(&obj, &source));
        source.insert("c".to_string(), 3);
        assert!(!conforms_to(&obj, &source));
    }

    #[test]
    fn test_eq_gt_gte_lt_lte() {
        assert!(eq(&3, &3));
        assert!(!eq(&3, &4));
        assert!(gt(5, 3));
        assert!(!gt(3, 5));
        assert!(gte(5, 5));
        assert!(!gte(4, 5));
        assert!(lt(3, 5));
        assert!(!lt(5, 3));
        assert!(lte(5, 5));
        assert!(!lte(6, 5));
    }

    #[test]
    fn test_is_array() {
        assert!(is_array(&[1, 2, 3]));
        assert!(is_array::<i32>(&[]));
    }

    #[test]
    fn test_is_boolean() {
        assert!(is_boolean(&true as &dyn Any));
        assert!(!is_boolean(&42 as &dyn Any));
    }

    #[test]
    fn test_is_date() {
        assert!(is_date(&SystemTime::now() as &dyn Any));
        assert!(!is_date(&42 as &dyn Any));
    }

    #[test]
    fn test_is_empty() {
        assert!(is_empty::<i32>(&[]));
        assert!(!is_empty(&[1]));
    }

    #[test]
    fn test_is_equal() {
        assert!(is_equal(&[1, 2], &[1, 2]));
        assert!(!is_equal(&[1], &[2]));
    }

    #[test]
    fn test_is_equal_with() {
        assert!(is_equal_with(&"abc", &"ABC", |a, b| a.to_lowercase() == b.to_lowercase()));
        assert!(!is_equal_with(&"abc", &"xyz", |a, b| a == b));
    }

    #[test]
    fn test_is_error() {
        assert!(is_error(&"error message".to_string() as &dyn Any));
        assert!(is_error(&std::io::Error::new(std::io::ErrorKind::Other, "err") as &dyn Any));
        assert!(!is_error(&42 as &dyn Any));
    }

    #[test]
    fn test_is_finite() {
        assert!(is_finite(1.0));
        assert!(!is_finite(f64::INFINITY));
        assert!(!is_finite(f64::NEG_INFINITY));
        assert!(!is_finite(f64::NAN));
    }

    #[test]
    fn test_is_function() {
        assert!(!is_function());
    }

    #[test]
    fn test_is_integer() {
        assert!(is_integer(3.0));
        assert!(!is_integer(3.5));
        assert!(!is_integer(f64::NAN));
    }

    #[test]
    fn test_is_length() {
        assert!(is_length(0));
        assert!(is_length(100));
        assert!(is_length(u32::MAX as usize));
        assert!(!is_length(u32::MAX as usize + 1));
    }

    #[test]
    fn test_is_map() {
        let m: HashMap<String, String> = HashMap::new();
        assert!(is_map(&m));
    }

    #[test]
    fn test_is_nan() {
        assert!(is_nan(f64::NAN));
        assert!(!is_nan(3.0));
    }

    #[test]
    fn test_is_nil_and_is_null() {
        let some_val: Option<i32> = Some(42);
        let none_val: Option<i32> = None;
        assert!(!is_nil(&some_val));
        assert!(is_nil(&none_val));
        assert!(!is_null(&some_val));
        assert!(is_null(&none_val));
    }

    #[test]
    fn test_is_number() {
        assert!(is_number(&42i32 as &dyn Any));
        assert!(is_number(&3.14f64 as &dyn Any));
        assert!(is_number(&100u64 as &dyn Any));
        assert!(!is_number(&"hello" as &dyn Any));
    }

    #[test]
    fn test_is_number_f64() {
        assert!(is_number_f64(3.14));
        assert!(is_number_f64(f64::NAN));
    }

    #[test]
    fn test_is_object() {
        assert!(is_object(&42));
        assert!(is_object(&"hello"));
    }

    #[test]
    fn test_is_plain_object() {
        assert!(is_plain_object(&42));
    }

    #[test]
    fn test_is_reg_exp() {
        assert!(!is_reg_exp(&"pattern" as &dyn Any));
    }

    #[test]
    fn test_is_safe_integer() {
        assert!(is_safe_integer(3.0));
        assert!(is_safe_integer(9007199254740991.0));
        assert!(!is_safe_integer(9007199254740992.0));
        assert!(!is_safe_integer(f64::NAN));
        assert!(!is_safe_integer(3.5));
    }

    #[test]
    fn test_is_set() {
        assert!(!is_set());
    }

    #[test]
    fn test_is_string() {
        assert!(is_string(&"hello".to_string() as &dyn Any));
        assert!(!is_string(&42 as &dyn Any));
    }

    #[test]
    fn test_is_symbol() {
        assert!(!is_symbol());
    }

    #[test]
    fn test_is_typed_array() {
        assert!(is_typed_array(&[1, 2]));
        assert!(is_typed_array::<i32>(&[]));
    }

    #[test]
    fn test_is_undefined() {
        assert!(!is_undefined());
    }

    #[test]
    fn test_is_weak_map() {
        assert!(!is_weak_map());
    }

    #[test]
    fn test_is_weak_set() {
        assert!(!is_weak_set());
    }

    #[test]
    fn test_to_array() {
        assert_eq!(to_array(&[1, 2, 3]), vec![1, 2, 3]);
        assert_eq!(to_array::<i32>(&[]), vec![]);
    }

    #[test]
    fn test_to_finite() {
        assert_eq!(to_finite(3.14), 3.14);
        assert_eq!(to_finite(f64::INFINITY), 0.0);
        assert_eq!(to_finite(f64::NAN), 0.0);
    }

    #[test]
    fn test_to_integer() {
        assert_eq!(to_integer(3.14), 3);
        assert_eq!(to_integer(-2.7), -2);
    }

    #[test]
    fn test_to_length() {
        assert_eq!(to_length(5), 5);
        assert_eq!(to_length(u32::MAX as usize + 1), u32::MAX as usize);
    }

    #[test]
    fn test_to_number() {
        assert_eq!(to_number("3.14"), 3.14);
        assert_eq!(to_number("abc"), 0.0);
    }

    #[test]
    fn test_to_plain_object() {
        assert_eq!(to_plain_object(42), 42);
    }

    #[test]
    fn test_to_safe_integer() {
        assert_eq!(to_safe_integer(3.14), 3);
        assert_eq!(to_safe_integer(1e20), 0);
        assert_eq!(to_safe_integer(-1e20), 0);
    }

    #[test]
    fn test_to_string_display() {
        assert_eq!(to_string(&42), "42");
        assert_eq!(to_string(&"hello".to_string()), "hello");
    }
}
